src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,375 @@
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext callContext) override {
+    auto methodName = kj::heapString(callContext.getParams().getMethodName());
+    auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+    bool caughtException = false;
+    // Try to execute the requested method.
+    try {
+      caughtException = co_await ctx.run(
+          [this, methodName=kj::mv(methodName), serializedArgs = kj::mv(serializedArgs),
+           entrypointName = entrypointName, &callContext] (Worker::Lock& lock) ->
+              kj::Promise<bool> {
+
+        bool exceptionThrown = false;
+        auto reader = callContext.initResults(capnp::MessageSize { 4, 1 }).initResult();
+        jsg::Lock& js = lock;
+
+        // We will try to get the function, if we can't we'll throw an error to the client.
+        auto maybeFn = tryGetFn(lock, entrypointName, ctx, methodName);
+        KJ_IF_SOME(e, maybeFn.tryGet<kj::String>()) {
+          exceptionThrown = true;
+          reader.setV8Serialized(_::serializeV8(js, js.error(e)));
+          co_return exceptionThrown;
+        }
+
+        auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+        kj::Maybe<SerializedResult> eventualResult;
+
+        // We have a function, so let's call it, get the result, and serialize it for RPC.
+        // We account for both the JS method throwing and our serialization attempt failing.
+        if (fn->IsAsyncFunction()) {
+          eventualResult = co_await prepareAsyncResult(js, fn, serializedArgs);
+        } else {
+          eventualResult = prepareSyncResult(js, fn, serializedArgs);
         }
-    ));
+        auto result = kj::mv(KJ_REQUIRE_NONNULL(eventualResult));
+        exceptionThrown = result.caughtException;
+
+        // Note that returning an exception from JS will not set `exceptionThrown`.
+        reader.setV8Serialized(kj::mv(result.data));
+        co_return exceptionThrown;
+      });
+    } catch(kj::Exception e) {
+      caughtException = true;
+      if (auto desc = e.getDescription();
+          !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
+        LOG_EXCEPTION("JsRpcTargetCall"_kj, e);
+      }
+    }
+    // Upon returning, we want to fulfill the callPromise so customEvent can continue executing.
+    // Recall that we inform the customEvent whether we completed cleanly, i.e. if we threw an
+    // exception.
+    KJ_DEFER(callFulfiller->fulfill(kj::mv(caughtException)););
+    co_return;
+  }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // The following names are registered by the Workers Runtime and cannot be called over RPC.
+  bool isRegisteredName(kj::StringPtr name) {
+    if (name == "constructor"_kjc ||
```

Comment:
As discussed before, we should block all methods inherited from `Object`, not just `constructor`. (And meanwhile if someone actually declares a method on their DO class called `constructor`, they should be allowed to call that. It's only the inherited version from `Object` that should be blocked.)

To implement this, I suppose you may need to manually implement the prototype walk. That is, you need to call `HasOwnProperty()` to check if the property exists on the object. If not, you call `GetPrototype()` to get the object's prototype, and then repeat `HasOwnProperty()` on that, and so on, until you get to the prototype for `Object`, which you should _not_ search. Only call `Get()` if one of these `HasOwnProperty()` calls returned true, otherwise behave like the method isn't present.

Unless @jasnell has a better idea...

Reply:
Just want to confirm something here, I have a handle to the object (class name `Actor`) and I'm testing `HasOwnProperty` with the following names: `[constructor, fetch, alarm, webSocketMessage, getString]`.

My DO class implements every name but for `webSocketMessage`.

I'm using `GetConstructorName()` to find out when I reach the `Object` prototype.
```
Actor hasOwn constructor? false
Actor hasOwn fetch? false
Actor hasOwn alarm? false
Actor hasOwn webSocketMessage? false 
Actor hasOwn getString? false

-- Actor Prototype (called GetPrototype of ^) --
Object hasOwn constructor? true
Object hasOwn fetch? true
Object hasOwn alarm? true
Object hasOwn webSocketMessage? false
Object hasOwn getString? true

-- Object Prototype (called GetPrototype of ^) --
Object hasOwn constructor? true
Object hasOwn fetch? false
Object hasOwn alarm? false
Object hasOwn webSocketMessage? false
Object hasOwn getString? false
```

It sounds like the last block is the one we shouldn't search (which has `constructor == true`, everything else false), but it's okay to call `Get('constructor')` on the second block prototype, right?

> And meanwhile if someone actually declares a method on their DO class called constructor, they should be allowed to call that. It's only the inherited version from Object that should be blocked.

I think I'm a bit confused by this, does this apply to all methods defined by the DO? What should be considered a "registered" method then (aside from properties of the `Object` prototype)?

Reply:
To rephrase: We should do the prototype walk and confirm the method is actually defined on the DO/its prototype (_not_ on `Object`s prototype), and then confirm its not one of _our_ registered methods (things we have hardcoded within the Workerd runtime like `fetch()`) -- or vice versa.

Reply:
If someone declares a `constructor` method on their DO then I think we should force that to be called with a modified name on the stub, e.g. `stub.constructor_()` ... similar to how we end up with stuff like `delete_()` on internal uses where the name conflicts with a keyword. It's a bit cumbersome but avoids the confusion.

> To implement this, I suppose you may need to manually implement the prototype walk. That is, you need to call HasOwnProperty() to check if the property exists on the object. If not, you call GetPrototype() to get the object's prototype, and then repeat HasOwnProperty() on that, and so on, until you get to the prototype for Object, which you should not search. Only call Get() if one of these HasOwnProperty() calls returned true, otherwise behave like the method isn't present.

You can simply call `Has(...)` to determine if the property exists anywhere in the prototype chain. https://v8docs.nodesource.com/node-20.3/db/d85/classv8_1_1_object.html#a23e79bb0b7a525fc6b36e3419a343833

Reply:
@irvinebroque do you have any thoughts here? As I understand it, the design doc suggests (under **"Reserved" handler called via RPC**) that calling methods that are handlers for features (like `alarm`, or `webSocketMessage`, `fetch`, possibly `constructor`) should throw.

> You can simply call `Has(...)` to determine if the property exists anywhere in the prototype chain. 

Doesn't that imply the `Object` prototype too, which we don't want to search?

Reply:
Sure but those are fairly easy to filter out by querying the same property off the `Object` prototype and checking for equality. Something equivalent to:

```
if (foo.Get("bar") === objectProto.Get("bar")) // then it's an Object prototype property
 ```

Reply:
What's the case where it is useful to allow someone to call a `constructor` method of a class via RPC? I get that this is possible in JS, but isn't this kind of a historical accident stemming from evolution of the language? Is safest thing to disallow?

Writing this out more — when in this lifecycle below does is the instance of the DO created, and the constructor called?

```js
const myCounter = env.COUNTER.get(id); // constructor not yet called
myCounter.foo(); // this actually creates the instance of the object, invoking constructor?
```

Reply:
@irvinebroque your example is right.

Also fwiw, attempting to call `stub.constructor()` on the client side throws `TypeError: Illegal constructor`. Since the client stub will never send `constructor` to the remote it doesn't matter if we check on the server side.

Reply:
Sorry, I realize this thing about "constructor" is extremely confusing.

There are two distinct concepts here:

1. A class's constructor, which is invoked when you create an instance of the class.
2. A regular method whose name happens to be "constructor", which would be invoked like `obj.constructor()`.

The second one does not normally appear in real code, but it's theoretically possible to declare, like:

To add confusion, the ES6 syntax for declaring (1) _looks like_ (2). That is, you declare a class's constructor using syntax that looks like you're defining a method called "constructor". However, in this context, the identifier "constructor" is magic and does not declare a regular method. Nevertheless, it is possible to create an object which has a method called `constructor`, e.g. by directly assigning the property.

```
class Foo {
  constructor() {
    // This is the constructor, called when you do `new Foo()`. This is NOT a
    // method called `constructor`, despite the syntax.

    this.constructor = () => {
      // This very silly function will be called when you do `foo.constructor()`
      // (on an existing instance of `Foo`).
    }
  }
}
```

To add another layer of confusion, assuming you do _not_ do shenanigans to create a method or property called `constructor` on your class, then there is by default a property called `constructor` inherited from `Object`. This property is a reference to the object's class itself, that is:

```
> new Foo().constructor === Foo
true
```

This property exists _whether or not_ your class explicitly declares a constructor.

So:
* We do not want to expose this `constructor` property that is inherited from the prototype of `Object`.
* However, if the object for some reason has a direct property called `constructor`, we can allow this to be invoked.
* The reason for this has nothing at all to do with constructors. The point here is that properties inherited from `Object` are properties that every object has whether it wants to or not. The application presumably does not intend for these to be exported over RPC. However, methods declared on the application's specific subclass are under the application's control. We are assuming any public method the application declares explicitly is intended to be exported over RPC.
* This all applies to _all properties inherited from `Object`_, e.g. all the methods listed in [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object). But only if the application does not override these methods with its own meanings.